### PR TITLE
Allow for "when" clause filtering for Notebook Toolbar Given ProviderId Changes

### PR DESCRIPTION
**Pull request by chlafreniere**
This allows for "when" clause filtering for the notebook toolbar; the notebook's providerId can change with the kernel (e.g. with the SQL kernel, the providerId is 'sql', but for the Python kernel, the providerId is 'Jupiter').

Currently, we have the Install Packages button, which is only valid for the 'jupyter' providerId. We want to disable it for any other providerId.

For the SQL taskbar, I needed a way to iterate through all of the items (in order to find the correct one to disable/enable it), so I added an items property. Please let me know if there's a better way to do this, or if you would prefer a different implementation/way to expose the items.

### CODE REVIEW

**Comment by anthonydresser** [comment]
Maybe i don't understand your code specifically but the actions that you pass into the action bar are passed as objects (i.e it doesn't instantiate it's own version of it) so if you disable the action you pass in it will show disabled. And you should have access to all the actions you pass to the bar since you pass them to it.

**Comment by chlafreniere** [comment]
@anthonydresser I see what you mean, but I have a minor problem. Disabling/enabling totally works as you suggested.

How do I determine if an action already has been added to the taskbar if I can't access the items, though? The problem is that we want to dynamically add (and later, probably remove) to the taskbar depending on a property in the notebook's model (which can change depending on which item is selected in another dropdown in the same taskbar). I'd probably need to have (yet another) array that keeps track of the items currently in the taskbar in the notebook component, and I'd really rather not have to do that. We already have 2 underlying arrays in the taskbar. Any ideas?

**Comment by anthonydresser** [comment]
Is there a problem with disabling a action even if it's not actually on the action bar?

I see here: https://github.com/Microsoft/azuredatastudio/pull/3712/files#diff-9395791fb352e01558cdccfd51f30a3bR499 you are also using it in order to avoid adding the same action multiple times. Everywhere else menu's are used in this way (with contributed actions) the menus are just rebuilt every time. It may seem inefficient but it happens so fast, there is no decernable difference to the user and it avoids having to iterate over all the actions in the action bar already.

**Comment by chlafreniere** [comment]
Ok.

File: src/sql/parts/notebook/models/notebookModel.ts [hunk 1]
```diff
@@ -559,7 +564,10 @@ export class NotebookModel extends Disposable implements INotebookModel {
 				let index = this.notebookManagers[i].sessionManager.specs.kernels.findIndex(kernel => kernel.name === kernelSpec.name);
```
Comment (kevcunnane) [inline src/sql/parts/notebook/models/notebookModel.ts]: Isn't this function really updateProviderId?

File: src/sql/parts/notebook/notebook.component.ts [hunk 2]
```diff
@@ -313,6 +318,16 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
 		this._changeRef.detectChanges();
 	}
 
+	private handleProviderIdChanged(providerId: string) {
+		// If there are any actions that were related to the previous provider,
+		// disable them in the actionBar
+		this._providerRelatedActions.forEach(action => {
+			action.enabled = false;
+		});
+		this.setContextKeyServiceWithProviderId(providerId);
```
Comment (kevcunnane) [inline src/sql/parts/notebook/notebook.component.ts]: Minor: maybe merge these two / create a wrapper call that is `updateProviderAndRefreshActions`

File: src/sql/parts/notebook/notebook.component.ts [hunk 3]
```diff
@@ -477,6 +476,40 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
...
+				this._providerRelatedActions.push(action);
+			} else {
+				this._providerRelatedActions[foundIndex].enabled = true;
+			}
+		}
+	}
+
+	private setContextKeyServiceWithProviderId(providerId: string) {
+		let provider = new RawContextKey<string>('providerId', providerId);
+		provider.bindTo(this.contextKeyService);
```
Comment (kevcunnane) [inline src/sql/parts/notebook/notebook.component.ts]: I think you can save the context key from this, and end up avoiding recreating a context key each time.

**Pull request merged**